### PR TITLE
add fieldmesh and make derivatives default to fieldmesh

### LINKED ISSUE #596: Derivatives - why global mesh?
Should derivatives not default to there own mesh to take derivatives?

Currently:
https://github.com/boutproject/BOUT-dev/blob/master/src/sys/derivs.cxx#L87

I think something like:
```
const Field3D DDY(const Field3D &f, CELL_LOC outloc, DIFF_METHOD method) {
  return f.getMesh()->indexDDY(f,outloc, method) / f.getMesh()->coordinates()->dy;
}
```
and
```
Mesh * Field3D::getMesh(){
  if (fieldmesh){
    return fieldmesh;
  } else {
    return mesh;
  }
}
```
I hope this helps allowing several meshes 👍

### PR DESCRIPTION
**Pull request by dschwoerer**
fixes #596

### CODE REVIEW

File: include/field2d.hxx [hunk 1]
```diff
@@ -98,15 +98,15 @@ class Field2D : public Field, public FieldData {
...
    */
-  int getNx() const {return nx;};
+  int getNx() const override {return nx;};
   /*!
    * Return the number of nx points
    */
-  int getNy() const {return ny;};
+  int getNy() const override {return ny;};
   /*!
    * Return the number of nx points
```
Comment (d7919) [inline include/field2d.hxx]: Could you fix the comments whilst you're here?

File: include/field3d.hxx [hunk 2]
```diff
@@ -212,15 +212,15 @@ class Field3D : public Field, public FieldData {
...
    */
-  int getNx() const {return nx;};
+  int getNx() const override {return nx;};
   /*!
    * Return the number of nx points
    */
-  int getNy() const {return ny;};
+  int getNy() const override {return ny;};
   /*!
    * Return the number of nx points
```
Comment (d7919) [inline include/field3d.hxx]: Could you fix the comments whilst you're here?

**Comment by d7919** [comment]
Do some operators, like `vddx`, require some extra checking to make sure that both fields involved are on compatible meshes?

**Comment by bendudson** [comment]
Yes, probably best to add ASSERT statements. Since it should have very little overhead, I think the checks should be done all the time (ASSERT0), though ASSERT1 may be ok.

**Comment by d7919** [comment]
I guess we need to define what we mean by two compatible meshes -- do they have to be exactly the same mesh, just have the same number of points in all/some directions, be convertible with interpolation or similar. For now I guess we could just force it to be the same mesh as the "safest" bet.

**Comment by bendudson** [comment]
I think ensuring that they are the exact same mesh is the best and safest thing. If (oh glorious day) we ever get to the point of having multiple meshes, it should probably be the physics model author who explicitly converts from one mesh to another.

**Comment by dschwoerer** [comment]
Now all derivatives are checked in the mesh.
If another mesh is implemented, that removes this restriction, it can easily remove the check 👍
Also now, results should be created with the right mesh.

**Comment by ZedThree** [comment]
I think a similar thing to the derivatives needs to be done in `F3D_OP_REAL` etc., as currently the result will use the global mesh. Basically, anywhere there is a temporary `Field3D` created to store a result, it needs to use the same mesh as the input arguments.

**Comment by dschwoerer** [comment]
Yes, I didn't wanted to fix everything in BOUT in one go - but rather start somewhere ...

On the specific issue with F3D_OP_REAL: that would be easier to fix with PR #458 :)
#458 checks already whether both fields are at the same position ...

Besides that: I would prefer open an issue for every case you find where it is not fixed (if you don't want to fix it your self)

This PR shouldn't cause any issues that haven't been there before.

**Comment by bendudson** [comment]
Fixing everything in BOUT(++) may be a bit too much to ask, and probably impossible.

**Pull request merged**